cloud-gateway-service/src/main/java/org/zowe/apiml/cloudgatewayservice/scheduled/GatewayScanJob.java

```diff
@@ -0,0 +1,94 @@
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import org.zowe.apiml.cloudgatewayservice.service.GatewayIndexService;
+import org.zowe.apiml.cloudgatewayservice.service.InstanceInfoService;
+import org.zowe.apiml.cloudgatewayservice.service.ServiceInfo;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Scheduled job to refresh registry of all registered gateways services.
+ * Behaviour of the job can  be configured by the following settings:
+ * <pre>
+ *   apiml:
+ *     cloudGateway:
+ *       cachePeriodSec: - default value 120 seconds
+ *       parallelismLevel:  - default value 20
+ *       clientKeystore: - default value null
+ *       clientKeystorePassword: - default value null
+ *       gatewayScanJobEnabled: - default value true
+ * </pre>
+ */
+@EnableScheduling
+@Slf4j
+@Component
+@ConditionalOnExpression("${apiml.cloudGateway.gatewayScanJobEnabled:true}")
+@RequiredArgsConstructor
+public class GatewayScanJob {
+    public static final String GATEWAY_SERVICE_ID = "GATEWAY";
+    private final GatewayIndexService gatewayIndexerService;
+    private final InstanceInfoService instanceInfoService;
+    @Value("${apiml.cloudGateway.parallelismLevel:20}")
+    private int parallelismLevel;
+
+    @Scheduled(initialDelay = 5000, fixedDelayString = "${apiml.cloudGateway.refresh-interval-ms:30000}")
+    public void startScanExternalGatewayJob() {
+        log.debug("Scan gateways job start");
+        doScanExternalGateway()
+                .subscribe();
+    }
+
+    /**
+     * reactive entry point  for the external gateways index refresh
+     */
+    protected Flux<List<ServiceInfo>> doScanExternalGateway() {
+
+        Mono<List<ServiceInstance>> registeredGateways = instanceInfoService.getServiceInstance(GATEWAY_SERVICE_ID);
+        Flux<ServiceInstance> serviceInstanceFlux = registeredGateways.flatMapMany(Flux::fromIterable);
+
+        return serviceInstanceFlux
+                .flatMap(gatewayIndexerService::indexGatewayServices, parallelismLevel);
+    }
+
+    /**
+     * Helper to echo state os the services caches. Will be removed later when REST api
+     */
+    @Scheduled(initialDelay = 10000, fixedDelayString = "5000")
+    void listCaches() {
+
+        Map<String, List<ServiceInfo>> fullState = gatewayIndexerService.listRegistry(null, null);
+
+        log.debug("Cache having {} apimlId records", fullState.keySet().size());
+        for (String apimlId : fullState.keySet()) {
+            List<ServiceInfo> servicesInfo = gatewayIndexerService.listRegistry(apimlId, null).get(apimlId);
+            log.debug("\t {}-{} : found {} external services", apimlId, apimlId, servicesInfo.size());
+        }
+
+        Map<String, List<ServiceInfo>> allSysviews = gatewayIndexerService.listRegistry(null, "bcm.sysview");
```

Comment:
please use some zowe serviceId

Reply:
this is just temporary sample code which showing filtering by the `apiId`.